Month 8 / step 2

Shiker2032 opened the pull request.

Shiker2032 committed:
Получение ингредиентов через хранилище

Shiker2032 committed:
Оформление заказа через редюсеры

Shiker2032 committed:
переключение табов скроллом

Shiker2032 committed:
начал делать dnd

Shiker2032 committed:
замена булки

Shiker2032 committed:
удаление ингредиентов

Shiker2032 committed:
подсчет стоимости

Shiker2032 committed:
подсчет стоимости через редюсеры

Shiker2032 committed:
dnd сортировка

Shiker2032 committed:
починил модалку деталей заказа

Shiker2032 committed:
Отображение счетчика

Shiker2032 committed:
Подготовил к проверке

gennady-bars commented on src/index.js:
```diff
@@ -0,0 +1,25 @@
+
+const enhancer = composeEnhancers(applyMiddleware(thunk));
+
+const store = createStore(rootReducer, enhancer);
```
**Можно лучше**


* `store` обычно создается в отдельном файле `store.js` и в папке `services`, чтобы не засорять `index.js`

gennady-bars commented on src/services/reducers.js:
```diff
@@ -0,0 +1,144 @@
+    case GET_INGREDIENTS: {
+      return [...state, ...action.data];
+    }
+    case "RESET_INGREDIENTS": {
```
` "RESET_INGREDIENTS"`

Это строка


Каждый тип экшена должен быть вынесен в константу в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const RESET_INGREDIENTS = 'RESET_INGREDIENTS'`

gennady-bars commented on src/services/actions.js:
```diff
@@ -0,0 +1,81 @@
+
+const getIngredients = () => (dispatch) => {
+  fetch(`${apiConfig.url}/ingredients`)
+    .then((res) => res.json())
```
`    .then((res) => res.json())
`

Любой ответ от сервера нужно проверять на `ok` и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.


У Вас есть функция `parseResponse`. Вот ее нужно использовать для проверки любого ответа от сервера

gennady-bars commented on src/services/actions.js:
```diff
@@ -0,0 +1,81 @@
+    .catch((er) => console.log(er));
+};
+
+const setCurrentIngredient = (currentIngredient) => {
```
Отличные экшен-криейтеры сделали

gennady-bars commented on src/services/actions.js:
```diff
@@ -0,0 +1,81 @@
+      modalHendler(true);
+      dispatch({ type: RESET_ORDER });
+      dispatch({ type: RESET_INGREDIENTS });
+      dispatch(getIngredients());
```
`      dispatch(getIngredients());
`


не нужно делать запрос за ингредиентами после получения номера заказа. Вы ингредиенты уже получили 1 раз при загрузке сайта

gennady-bars commented on src/services/actions.js:
```diff
@@ -0,0 +1,81 @@
+      dispatch(setOrderNumber(json.order.number));
+      modalHendler(true);
+      dispatch({ type: RESET_ORDER });
+      dispatch({ type: RESET_INGREDIENTS });
```
`dispatch({ type: RESET_INGREDIENTS });`


Вы не те ингредиенты очищаете. Вы хотели очищать заказ, а очищаете все ингредиенты проекта


Нужно просто сам заказ обнулять


А цифры над ингредиентами тоже нужно обнулять, если очищаете заказ (но не путем очистки всех ингредиентов проекта)

gennady-bars commented on src/components/App/App.js:
```diff
@@ -1,36 +1,42 @@
     setIsIngredientsDetailsOpened(false);
   };
 
   const handleEscKeydownIngredientModal = (event) => {
```
Все лишние обработчики `Escape` нужно удалить из проекта. Уже есть универсальный внутри `Modal`. Именно он работает правильно

gennady-bars commented on src/components/App/App.js:
```diff
@@ -1,36 +1,42 @@
+    event.key === "Escape" && handleCloseIngredientModal();
   };
 
   const handleEscKeydownOrderModal = (event) => {
```
Все лишние обработчики `Escape` нужно удалить из проекта. Уже есть универсальный внутри `Modal`. Именно он работает правильно

gennady-bars commented on src/components/App/App.js:
```diff
@@ -42,45 +48,21 @@ function App() {
+      <DndProvider backend={HTML5Backend}>
         <main className={styles.app__flexComponents}>
           <BurgerIngredients
             ingredients={ingredients}
```
`ingredients={ingredients}` нужно удалить из пропсов `BurgerIngredients` и `BurgerConstructor`. Все теперь берется из редакса

gennady-bars commented on src/components/burger-constructor/BurgerConstructor.js:
```diff
@@ -0,0 +1,156 @@
+            order.map((ingredient, idx) => {
+              if (ingredient.price !== 0 && ingredient.type !== "bun") {
+                return (
+                  <li key={ingredient._id}>
```
` key={ingredient._id}`

для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время сброса в зону `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. 


А в `key` будет уже просто **key={item.uuid}**

gennady-bars commented on src/components/Burger-ingredients/IngredientsItem.js:
```diff
@@ -0,0 +1,65 @@
+
+IngredientsItem.propTypes = {
+  ingredient: PropTypes.object,
+  Onclick: PropTypes.func,
```
`Onclick` нет у Вас. Есть `onClick`

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная, но есть некоторые недочеты:



* проект не запускается. Скрин ошибки Скрин http://joxi.ru/BA0YjEvS19DKNr . Я исправил вручную название файла на `header.module.css` и все запустилось. 
* Видимо, кэш гита не отслеживает регистр файлов. Для гита без разницы, с маленькой буквы или с большой сделан файл. Он считает их одинаковыми. Попробуйте очистить кэш гита. Посоветуйтесь с наставником или посмотрите по этой ссылке, как это сделать https://coderoad.ru/41863484/%D0%9E%D1%87%D0%B8%D1%81%D1%82%D0%B8%D1%82%D1%8C-%D0%BB%D0%BE%D0%BA%D0%B0%D0%BB%D1%8C%D0%BD%D1%8B%D0%B9-%D0%BA%D1%8D%D1%88-git
* Каждый тип экшена должен быть вынесен в константу в папке `actions`, а потом его нужно использовать везде в коде. Это требуется для того, чтобы было одно место, где его можно изменить, подкорректировать, и не было шанса  ошибиться в написании одной и той же строки в 2-3х местах. `export const RESET_INGREDIENTS = 'RESET_INGREDIENTS'`
* Не могу добавить 2 и более одинаковых ингредиента, хотя счетчик растет. Скрин http://joxi.ru/EA4YWBvSXjEGYr 
* Неправильно работает логика: по заданию нужно сделать логику перетаскивания ингредиентов внутри заказа (вверх и вниз), а сейчас что-то непонятное происходит. А если добавить одинаковые ингредиенты и один другой, то вообще ничего не переносится
* У Вас есть функция `parseResponse`. Вот ее нужно использовать для проверки любого ответа от сервера 
* не нужно делать запрос за ингредиентами после получения номера заказа. Вы ингредиенты уже получили 1 раз при загрузке сайта
* Вы не те ингредиенты очищаете. Вы хотели очищать заказ, а очищаете все ингредиенты проекта
* Все лишние обработчики `Escape` нужно удалить из проекта. Уже есть универсальный внутри `Modal`. Именно он работает правильно
* `ingredients={ingredients}` нужно удалить из пропсов `BurgerIngredients` и `BurgerConstructor`. Все теперь берется из редакса
* для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время сброса в зону `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={item.uuid}**
* Нужно посмотреть все компоненты и типизировать их `пропсы`, если где-то пропустили. 
* Всем элементам `props`, которые являются обязательными, нужно добавлять `isRequired` при типизации с `propTypes`. Это касается самих массивов и объектов. В самом конце нужно указывать `isRequired`, если массив сам обязателен в `props`





**Можно лучше**


* `store` обычно создается в отдельном файле `store.js` и в папке `services`, чтобы не засорять `index.js`




Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Shiker2032 committed:
очистил кэш

Shiker2032 committed:
Рефакторинг

Отображение одинаковых элементов в конструкторе
Уникальный ключ для элементов
Запросы обрабатываются через parse
Store в services
Сброс заказа
Сброс счетчика ингредиентов
Лишние обработчики Esc
Лишние пропсы ингредиентов

Shiker2032 committed:
Удалить header

Shiker2032 committed:
Create header.module.css

Shiker2032 committed:
Шаманство с гитом

Shiker2032 committed:
показать гиту изменения

Shiker2032 committed:
починил удаление ингредиентов

Shiker2032 committed:
Починил счетчики

gennady-bars commented on src/services/reducers.js:
```diff
@@ -13,14 +14,33 @@ import {
-    case "RESET_INGREDIENTS": {
-      return (state = [{}]);
+    case RESET_INGREDIENTS: {
+      const arr = state.forEach((el) => {
```
` const arr = state.forEach((el) => {`

* в консоли пойманная ошибка после очистки заказа. Скрин http://joxi.ru/l2Zqb96HlOoJwm 



вот это не возвращает ничего в `arr`, поэтому `undefined` получается


Нужно использовать `map`, чтобы сохранить все данные об ингредиентах, но изменить только `amount`


```js

    case RESET_INGREDIENTS: {
      const arr = [...state].map((el) => {
        el.amount = 0;
        return el
      });
      return arr;
    }


```

gennady-bars commented on src/services/reducers.js:
```diff
@@ -41,9 +61,9 @@ const currentIngredientReducer = (state = { price: 0 }, action) => {
-      ingredient.amount = 1;
-      return [...state, ingredient];
+      const ingredeint = { ...action.ingredient };
+      ingredeint.uid = uuidv4();
```
`      ingredeint.uid = uuidv4();
`

Редьюсер должен быть чистой функцией.  Генерировать `uuid` в самом редьюсере нельзя. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в `action creator`). И созданный uuid должен приходить в редьюсер в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста.

gennady-bars commented on src/services/reducers.js:
```diff
@@ -13,14 +14,33 @@ import {
 import { GET_INGREDIENTS, SET_CURRENT_INGREDIENT, SET_ORDER } from "./types";
+import { v4 as uuidv4 } from "uuid";
 
 const ingredientsReducer = (state = [{}], action) => {
```
**Можно лучше**


Вам нужно сделать `initialState` для всех редьюсеров, чтобы все логично обрабатывать


```js

const ingredientsInitialState = {
  ingredients: []
}

const ingredientsReducer = (state = ingredientsInitialState, action) => {

```


а потом стейт обновлять нужно вот так везде:


```js

const ingredientsInitialState = {
  ingredients: []
}

const ingredientsReducer = (state = ingredientsInitialState, action) => {
  switch (action.type) {
    case GET_INGREDIENTS: {
      return {
        ...state, 
        ingredients: action.data
      }
    }


```


и так нужно со всеми стейтами сделать, а то что-то сложно приходится Вам изменять стейт

gennady-bars commented on src/components/Burger-constructor/BurgerConstructor.js:
```diff
@@ -94,7 +98,7 @@ const BurgerConstructor = ({ onClick }) => {
               if (ingredient.price !== 0 && ingredient.type !== "bun") {
                 return (
-                  <li key={ingredient._id}>
+                  <li key={uuidv4()}>
```
`key={uuidv4()`


в `key` нужно указывать только 1 раз созданный уникальный `uuid` ингредиента, иначе Вы сейчас меняете его при каждой перерисовке, что приводит к тому, что Реакт не может уследить за перемещением компонентов, и это может сломать логику `DnD`


А в `key` будет уже просто **key={item.uuid}**

gennady-bars reviewed (commented):
Еще немного и все будет отлично:

* в консоли ошибка типизации, когда начинаешь собирать заказ. Скрин http://joxi.ru/Dr8YZBESKj77jm 
* в консоли пойманная ошибка после очистки заказа. Скрин http://joxi.ru/l2Zqb96HlOoJwm 
* Редьюсер должен быть чистой функцией.  Генерировать `uuid` в самом редьюсере нельзя. Это делает редьюсер нечистой функцией. Нужно генерировать uuid при создании экшена (желательно, в `action creator`). И созданный uuid должен приходить в редьюсер в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста.
* в `key` нужно указывать только 1 раз созданный уникальный `uuid` ингредиента, иначе Вы сейчас меняете его при каждой перерисовке, что приводит к тому, что Реакт не может уследить за перемещением компонентов, и это может сломать логику `DnD`

Shiker2032 committed:
Исправления

Исправил проптайп
Исправил очистку ингредиентов
Установка заказа через генератор экшона
Фиксированный uid для элемента

Shiker2032 committed:
редюсеры с начальным состоянием

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
